nixos/mailpit: init

Ma27 opened the pull request:
## Description of changes
Simple module to configure a systemd service for mailpit.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
nixos/mailpit: init

Simple module to configure a systemd service for mailpit.

leona-ya commented on nixos/modules/services/mail/mailpit.nix:
```diff
@@ -0,0 +1,103 @@
+          ]
+        );
+        options = {
+          db-file = mkOption {
```
This seems to be outdated, but as nixpkgs has a quite old version of mailpit it seems correct for now..

Ma27 replied:
Oof I already wondered why how I came up with `database` while hacking on that module 🤦 
I guess I'll do an update of the software then, thanks for catching this.

RafaelKr replied:
FYI I already opened https://github.com/NixOS/nixpkgs/pull/324302 for automatic update PRs via @r-ryantm 
Still waiting to get it merged.

Btw I also initiated that the mailpit documentation shows the version in which specific options were added. See https://github.com/axllent/mailpit/issues/328 for more info.

RafaelKr commented:
May I suggest to make it possible to configure multiple servers? My use case is to have multiple environments (dev, staging, production) on one host and I like to have an instance for each.

I already have built something for a local project but didn't have time to work it out with all options and create a PR for it.

But here's my implementation: https://gist.github.com/RafaelKr/e9ac5850da8b674a92f0bf4bb5bb9b7f

Commit:
nixos/mailpit: allow multiple instances

Now it's possible to start multiple mailpit instances - for e.g.
multiple testing environments - on the same machine:

    {
      services.mailpit.instances = {
        dev = { /* ... */ };
        staging = { /* ... */ };
      };
    }

The simplest way to start a single instance is by declaring

    services.mailpit.instances.default = {};

Ma27 commented:
ping @leona-ya @RafaelKr 
Just rebased against latest master.
PR should be ready to merge I guess? :)

RafaelKr commented:
I think mailpit listens on `0.0.0.0` by default to make it easily usable inside docker. For usage with nix I would suggest switching to the more secure `127.0.0.1` by default. What do you think?
Besides from that it LGTM! :)

Maybe we want to wait for the merge of #334352 and rebase again before merging this PR.

RafaelKr commented on nixos/modules/services/mail/mailpit.nix:
```diff
@@ -0,0 +1,106 @@
+            '';
+          };
+          listen = mkOption {
+            default = "0.0.0.0:8025";
```
Maybe default to `127.0.0.1:8025`. See https://github.com/NixOS/nixpkgs/pull/329657#issuecomment-2288699086

RafaelKr commented on nixos/modules/services/mail/mailpit.nix:
```diff
@@ -0,0 +1,106 @@
+            '';
+          };
+          smtp = mkOption {
+            default = "0.0.0.0:1025";
```
Maybe default to `127.0.0.1:1025`. See https://github.com/NixOS/nixpkgs/pull/329657#issuecomment-2288699086

Ma27 commented:
> Maybe we want to wait for the merge of #334352 and rebase again before merging this PR.

This shouldn't give a conflict: `source.nix` isn't touched anymore in this PR.

Ma27 committed:
nixos/mailpit: fix defaults for listen/smtp

RafaelKr commented:
Let's go! 🚀 :)
